DylibBinder/DBAssociatedType.cs

```diff
@@ -0,0 +1,72 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using SwiftReflector;
+
+namespace DylibBinder {
+	internal class DBAssociatedType {
+		public string Name { get; }
+
+		public DBAssociatedType (string name)
+		{
+			if (string.IsNullOrEmpty (name))
+				throw new ArgumentException ("DBAssociateType name cannot be null or empty");
+			Name = name;
+		}
+	}
+
+	internal class DBAssociatedTypes {
+		public HashSet<DBAssociatedType> AssociatedTypeCollection { get; } = new HashSet<DBAssociatedType> (new DBAssociatedTypeComparer ());
+
+		public DBAssociatedTypes ()
+		{
+		}
+
+		public DBAssociatedTypes (DBTypeDeclaration typeDeclaration)
+		{
+			AssociatedTypeCollection.UnionWith (typeDeclaration.GetAssociatedTypes ());
+		}
+	}
+
+	// interface for having an AssociatedTypes Property
+	internal interface IAssociatedTypes {
+		public DBAssociatedTypes AssociatedTypes { get; }
+	}
+
+	internal static class DBAssociatedTypesExtensions {
+		public static List<DBAssociatedType> GetAssociatedTypes (this SwiftBaseFunctionType signature)
+			=> signature.EachParameter.SelectMany (t => AssociatedTypesSwitch (t)).ToList ();
+
+		public static List<DBAssociatedType> GetAssociatedTypes (this SwiftType swiftType)
+			=> AssociatedTypesSwitch (swiftType);
+
+		static List<DBAssociatedType> AssociatedTypesSwitch (object type) => type switch {
+			SwiftGenericArgReferenceType genericArgType => GetAssociatedTypes (genericArgType),
+			SwiftFunctionType funcType => GetAssociatedTypes (funcType),
+			_ => new List<DBAssociatedType> ()
+		};
+
+		static List<DBAssociatedType> GetAssociatedTypes (SwiftFunctionType funcType)
+			=> funcType.Parameters is SwiftGenericArgReferenceType genericArgType ? GetAssociatedTypes (genericArgType) : new List<DBAssociatedType> ();
+
+		static List<DBAssociatedType> GetAssociatedTypes (SwiftGenericArgReferenceType genericArgType)
+			=> genericArgType.AssociatedTypePath.Select (t => new DBAssociatedType (t)).ToList ();
+
+		public static List<DBAssociatedType> GetChildrenAssociatedTypes<T> (this IEnumerable<T> iEnumerableInstance)
+			=> iEnumerableInstance.Where (t => t is IAssociatedTypes).Cast<IAssociatedTypes> ().SelectMany (t => t.AssociatedTypes.AssociatedTypeCollection).ToList ();
+	}
+
+	internal class DBAssociatedTypeComparer : EqualityComparer<DBAssociatedType> {
+		public override bool Equals (DBAssociatedType at1, DBAssociatedType at2)
+		{
+			if (at1 == null && at2 == null)
+				return true;
+			else if (at1 == null || at2 == null)
+				return false;
+			return at1.Name == at2.Name;
+		}
+
+		public override int GetHashCode (DBAssociatedType at)
+			=> at == null ? "null".GetHashCode () : at.Name.GetHashCode ();
```

Comment:
If you get null, there is not need to do that. GetHashCode needs to return always the same value for the same object, that is all. You could have done:

```
if (at == null)
   return 42;
return at.Name.GetHashCode ();
```
GetHashCode is not guaranteed to return unique hash codes as per MSD docs:

> Two objects that are equal return hash codes that are equal. However, the reverse is not true: equal hash codes do not imply object equality, because different (unequal) objects can have identical hash codes. Furthermore, .NET does not guarantee the default implementation of the GetHashCode method, and the value this method returns may differ between .NET implementations, such as different versions of .NET Framework and .NET Core, and platforms, such as 32-bit and 64-bit platforms. For these reasons, do not use the default implementation of this method as a unique object identifier for hashing purposes. Two consequences follow from this:
> You should not assume that equal hash codes imply object equality.
> You should never persist or use a hash code outside the application domain in which it was created, because the same object may hash across application domains, processes, and platforms.

That is from: https://docs.microsoft.com/en-us/dotnet/api/system.object.gethashcode?view=net-5.0#remarks

Reply:
Makes sense! Thanks!